### ISSUE TITLE
Sort candidates alphabetically for non-list posts

### ISSUE THREAD

**Issue by jf1** [mysociety/yournextrepresentative#812]
As discussed with @mhl, for posts with individual candidates rather than a party list.
eg https://candidates.democracyclub.org.uk/election/mayor.salford.2016-05-05/post/mayor:salford/mayor-of-salford

DC has no surname field so will need to sort by the last word of the candidate's full name field.

**Comment by wfdd** [comment]
@jf1 @mhl

Shouldn't this have been merged in the DC repo instead?  Lots of languages have compound surnames, including Greek.

Also, if line 210 was removed because it's redundant, I don't think it is.  If I'm not mistaken, line 234 will sort the parties according to the last name of the first person on the list of each party.

**Comment by wfdd** [comment]
Never mind, it is actually redundant in practice - though this isn't gonna alphabetise unranked candidates in party lists.

**Comment by mhl** [comment]
@wfdd Yeah, I think it it redundant. (This function is hard to read, I'm afraid - it really needs some refactoring, or at the least we should use namedtuple instead of tuple here.)

To alphabetise unranked candidates in party lists, I think this will work (although obviously with some reformatting to make it clearer):

```diff
diff --git a/candidates/views/helpers.py b/candidates/views/helpers.py
index 4658a6a..8815b03 100644
--- a/candidates/views/helpers.py
+++ b/candidates/views/helpers.py
@@ -200,7 +200,7 @@ def group_candidates_by_party(election_data, candidacies, party_list=True, max_p
         total_count = len(people_list)
         if election_data.party_lists_in_use:
             # sort by party list position
-            people_list.sort(key=lambda p: ( p[0] is None, p[0] ))
+            people_list.sort(key=lambda p: ( p[0] is None, p[0], p[1].name.split()[-1] ))
             # only return the configured maximum number of people
             # for a party list
             if max_people and len(people_list) > max_people:
```

On the broader point, I know this isn't a great solution, but the current behaviour was to sort on the `family_name` field, which was empty in every deployment of YNR that I have access to, so there was no useful sorting being done as it was. So I thought this would be an incremental improvement, if not my preferred solution. Had you been populating `family_name` or `sort_name` in your installation? If so, sorry, this will have made things worse, and I'll figure out something else.

My preferred solution is to make sure that the Popolo `sort_name` is populated, and use that instead.

**Comment by wfdd** [comment]
Yes, we have `sort_name` but I don't think we're affected by this commit on closer look, since we use party lists.  (I'd added a third item in the the sort function for `sort_name` like you've done in the diff.)  It'd probably be good to swap in `sort_name` at some point in the future.

**Comment by jf1** [comment]
A minor point, but after sorting alphabetically on the last word, it's not then doing any further sorting by the first word. Example from [this ward](https://candidates.democracyclub.org.uk/election/local.trafford.2016-05-05/post/MTW:E05000831/hale-central-ward).
![image](https://cloud.githubusercontent.com/assets/4711363/14569842/080667e8-0339-11e6-9278-ba90716f4443.png)
(I did check the SoPN to confirm they're not listed at the same address - that'd be fireworks over the dinner table!)

**Comment by jf1** [comment]
Another post whose candidates aren't sorted alphabetically, making checking then locking harder than it needs to be. The SoNP is sorted alphabetically, and although there can be more than one candidate per party it's an STV system so individual candidates aren't ordered within party lists.
https://candidates.democracyclub.org.uk/election/nia.2016-05-05/post/osni_oid:NIE-3/member-of-the-legislative-assembly-for-east-antrim

**Comment by mhl** [comment]
@jf1 Yes, I know, this is a annoying to me too, sorry. I'll reopen the ticket. What's happening is that the candidates are being grouped by party, and then the parties groups are being ordered by the last name of the first person in each party group, which isn't very helpful to anyone! I think in the non-party list case, only one of these makes sense:

1. Group by party, and order by party name first, then candidate's name
2. Don't group by party, and order all candidates by candidate's name

Obviously we want the second in the UK case to make checking SOPNs easier, though there may be cases internationally where the second would be preferred. As a result, I think a pull request to the DC fork would make most sense to get behaviour 2 in the non-party list case.

**Comment by jf1** [comment]
Thought it was something like that but couldn't deduce why the parties weren't ordered alphabetically by their party name. Agree the existing behaviour is illogical and without useful purpose.

In the NI example scenario 1. would rarely be useful, but I can see the value in providing it as an option for occasional use. Also think there's value in providing vice-versa for party-list posts. Could each page be capable of both 1. and 2. with party-list posts defaulting to 1. and others defaulting to 2. - and with both of them having a link to sort by the alternative method?

Suggestions for friendly names:
1. Sort grouped by party
2. Sort by candidate name

**Comment by wfdd** [comment]
Importing the sort function form `election_extra` might be an idea?  Then we can revert to a saner default and everybody can use whatever sorting algorithm they like.  Though we're also gonna have to 'unwrap' the list of candidates for non-party lists.

**Comment by mhl** [comment]
As a quick fix for the UK case, I've made this pull request against the Democracy Club fork: https://github.com/DemocracyClub/yournextrepresentative/pull/18

**Comment by jf1** [comment]
Thanks Mark. 

I spotted another post that's not sorted by candidate surname, and this time it's not sorted using the criteria you described above, so is it worth checking that your pull request will fix [this post](https://candidates.democracyclub.org.uk/election/local.lincoln.2016-05-05/post/NODATA:E07000138-abbey/abbey-ward) too?

@symroe the NODATA part of [this URL](https://candidates.democracyclub.org.uk/election/local.lincoln.2016-05-05/post/NODATA:E07000138-abbey/abbey-ward) should be "DIW" and all Lincoln wards listed [here](https://candidates.democracyclub.org.uk/election/local.lincoln.2016-05-05/constituencies) are similarly affected, perhaps because they've had [boundary changes](http://2016wards.mysociety.org/postcode/LN25EH).

**Comment by mhl** [comment]
@jf1 It should fix that one too. When I said:

> What's happening is that the candidates are being grouped by party, and then the parties groups are being ordered by the last name of the first person in each party group, which isn't very helpful to anyone!

It may not have been clear that the candidates within the party groups aren't being sorted, (since 45eb39eb21dfb65e93, sigh) so "the first person in each party group" isn't necessarily the candidate of the party that's earliest in the alphabet.
